llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp

```diff
@@ -3510,6 +3524,12 @@ Value *InstCombinerImpl::foldBooleanAndOr(Value *LHS, Value *RHS,
       if (Value *Res = foldAndOrOfICmps(LHSCmp, RHSCmp, I, IsAnd, IsLogical))
         return Res;
 
+  /// Try to fold (icmp(A & B) ==/!= C) &/| (icmp(A & D) ==/!= E)
+  /// into a single (icmp(A & X) ==/!= Y).
+  if (Value *V = foldLogOpOfMaskedICmps(LHS, RHS, IsAnd, IsLogical, Builder,
+                                        SQ.getWithInstruction(&I)))
+    return V;
```

Comment:
Maybe keep the original place of the call in this commit so this change can be NFC. Would simplify review.